src/reader.rs

```diff
@@ -20,13 +21,22 @@ use std::fmt::{Display, Formatter};
 
 use crate::types::Str;
 /// Configures and constructs new instances of [Reader].
-pub struct ReaderBuilder {}
+pub struct ReaderBuilder {
+    // This will be set to to `Some` catalog when the reader builder is created with catalog information.
+    // Default value for this field will be set to `None`.
+    catalog: Option<Box<dyn Catalog>>,
```

Comment:
I think we can use [the `EmptyCatalog` type](https://github.com/amazon-ion/ion-rust/blob/ef641f48cdbc4da2ba947ae02edd540bdd12ef98/src/catalog.rs#L71) you wrote as the default instead of `None`. Because `EmptyCatalog` is a zero-sized type, `Box::new(EmptyCatalog)` won't actually allocate.

Reply:
I tried using that but the issue with using `EmptyCatalog` was we can't distinguish between a reader with an actually empty catalog vs a reader without a catalog. Which is why I chose the `Option` here. This distinction will be useful for creating the reader from builder in `make_*_reader()`.

Reply:
Can't you replace:
```rust
    fn make_binary_reader<'a, I: 'a + IonDataSource>(self, data: I) -> IonResult<Reader<'a>> {
        let raw_reader = Box::new(BlockingRawBinaryReader::new(data)?);
        if let Some(catalog) = self.catalog {
            return Ok(Reader::with_catalog(raw_reader, catalog));
        }
        Ok(Reader::new(raw_reader))
    }
```
with:
```rust
    fn make_binary_reader<'a, I: 'a + IonDataSource>(self, data: I) -> IonResult<Reader<'a>> {
        let raw_reader = Box::new(BlockingRawBinaryReader::new(data)?);
        Ok(Reader::with_catalog(raw_reader, self.catalog))
    }
```
or am I missing something?

Reply:
Thats right. My bad. I thought this distinction was required at system reader level but even that uses empty catalog as default.